src/raw/mod.rs

```diff
@@ -0,0 +1,1634 @@
+    /// ```
+    /// # use bson::raw::RawDocument;
+    /// use bson::doc;
+    ///
+    /// let doc = RawDocument::from_document(&doc!{});
+    /// assert_eq!(doc.into_inner(), b"\x05\x00\x00\x00\x00".to_vec());
+    /// ```
+    pub fn into_inner(self) -> Vec<u8> {
+        self.data.to_vec()
+    }
+}
+
+impl TryFrom<RawDocument> for Document {
+    type Error = RawError;
+
+    fn try_from(raw: RawDocument) -> RawResult<Document> {
+        Document::try_from(raw.as_ref())
+    }
+}
+
+impl<'a> IntoIterator for &'a RawDocument {
+    type IntoIter = RawDocumentIter<'a>;
+    type Item = RawResult<(&'a str, RawBson<'a>)>;
+
+    fn into_iter(self) -> RawDocumentIter<'a> {
+        RawDocumentIter {
+            doc: &self,
+            offset: 4,
+        }
+    }
+}
+
+impl AsRef<RawDocumentRef> for RawDocument {
+    fn as_ref(&self) -> &RawDocumentRef {
+        RawDocumentRef::new_unchecked(&self.data)
+    }
+}
+
+impl Borrow<RawDocumentRef> for RawDocument {
+    fn borrow(&self) -> &RawDocumentRef {
+        &*self
+    }
+}
+
+impl ToOwned for RawDocumentRef {
+    type Owned = RawDocument;
+
+    fn to_owned(&self) -> Self::Owned {
+        self.to_raw_document()
+    }
+}
+
+/// A BSON document referencing raw bytes stored elsewhere. This can be created from a
+/// [RawDocument] or any type that contains valid BSON data, and can be referenced as a `[u8]`,
+/// including static binary literals, [Vec<u8>](std::vec::Vec), or arrays.
+///
+/// Accessing elements within a `RawDocumentRef` is similar to element access in [bson::Document],
+/// but because the contents are parsed during iteration, instead of at creation time, format errors
+/// can happen at any time during use.
+///
+/// Iterating over a RawDocumentRef yields either an error or a key-value pair that borrows from the
+/// original document without making any additional allocations.
+
+/// ```
+/// # use bson::raw::{Doc, RawError};
+/// let doc = RawDocumentRef::new(b"\x13\x00\x00\x00\x02hi\x00\x06\x00\x00\x00y'all\x00\x00")?;
+/// let mut iter = doc.into_iter();
+/// let (key, value) = iter.next().unwrap()?;
+/// assert_eq!(key, "hi");
+/// assert_eq!(value.as_str(), Ok("y'all"));
+/// assert!(iter.next().is_none());
+/// # Ok::<(), RawError>(())
+/// ```
+///
+/// Individual elements can be accessed using [`RawDocumentRef::get`](RawDocumentRef::get) or any of
+/// the type-specific getters, such as
+/// [`RawDocumentRef::get_object_id`](RawDocumentRef::get_object_id) or [`RawDocumentRef::
+/// get_str`](RawDocumentRef::get_str). Note that accessing elements is an O(N) operation, as it
+/// requires iterating through the document from the beginning to find the requested key.
+///
+/// ```
+/// # use bson::raw::{RawDocument, RawError};
+/// let doc = RawDocument::new(b"\x13\x00\x00\x00\x02hi\x00\x06\x00\x00\x00y'all\x00\x00".to_vec())?;
+/// assert_eq!(doc.get_str("hi")?, Some("y'all"));
+/// # Ok::<(), RawError>(())
+/// ```
+#[derive(Debug)]
+pub struct RawDocumentRef {
```

Comment:
We'll probably also want a mutable version of this type to support appending to subdocuments without copying, ditto for arrays. I don't think there's a way for us to have an impl block conditional on whether the underlying reference is mutable or not, which would let us have a single type.

Reply:
That won't work, since the subdocument is part of a larger Vec, so even if it were made mutable, the mutability would be strictly limited to changing values that are fixed size.  For those kinds of modifications, you really do need to parse the document into a more flexible type, and take the performance hit.